src/pconn.h

```diff
@@ -81,9 +81,9 @@ class IdleConnList: public hash_link, private IndependentRunner
     Comm::ConnectionPointer *theList_;
 
     /// Number of entries theList can currently hold without re-allocating (capacity).
-    int capacity_;
+    size_t capacity_;
     ///< Number of in-use entries in theList
-    int size_;
+    size_t size_;
```

Comment:
Prior to this change, calling IdleConnList::pop() on an empty list would result in a nil pointer getting returned. Now, I would expect a segmentation fault or worse. The method is called effectively unconditionally from Adaptation::Icap::ServiceRep::getIdleConnection() (at least).

```c++
IdleConnList::pop()
{
    for (int i=size_-1; i>=0; --i) ...
}
```

If possible, undo all pconn-related changes so that you can make progress without rewriting that code.

How do we know that there are no other, similar bugs in this PR changes? Perhaps its scope should be reduced to assist with manual validation?

Reply:
This is coming from initial PR comments https://github.com/squid-cache/squid/pull/1118/files#r941969374
changing type of size_ leads to other variables changes that are compared or else with size_

As for testing, I do compiling in my context but, for the rest, the CI will hopefully catch most of it.

Reply:
> This is coming from initial PR comments https://github.com/squid-cache/squid/pull/1118/files#r941969374

I see. It is not clear (to me) what specifically triggered the changes that triggered those (correct in principle but insufficiently insightful) initial PR comments. My current recommendation stands: If possible, undo all pconn-related changes, so that you can make progress here without rewriting that code. If not possible, let's discuss why it is not possible.

> As for testing, ... the CI will hopefully catch most of it.

It will not. Our test coverage is not ye sufficient to rely on CI for such changes, unfortunately. For example, ICAP code (that PR pconn changes probably break) is not exercised by current CI at all IIRC.

Reply:
Reverting done but not compiling anymore
```
g++ -std=c++17 -DHAVE_CONFIG_H -DDEFAULT_CONFIG_FILE=\"/etc/squid/squid.conf\" -DDEFAULT_SQUID_DATA_DIR=\"/usr/share/squid\" -DDEFAULT_SQUID_CONFIG_DIR=\"/etc/squid\"   -I.. -I../include -I../lib -I../src -I../include  -isystem /usr/include/mit-krb5   -I../src  -isystem /usr/include/mit-krb5  -Wdate-time -D_FORTIFY_SOURCE=3 -Wall -Wextra -Wimplicit-fallthrough=5 -Wpointer-arith -Wwrite-strings -Wcomments -Wshadow -Wmissing-declarations -Woverloaded-virtual -Werror -pipe -D_REENTRANT -I/usr/include/p11-kit-1    -g -O2 -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -ffile-prefix-map=/var/cache/build/squid/squid-6.12=. -flto=auto -ffat-lto-objects -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -fdebug-prefix-map=/var/cache/build/squid/squid-6.12=/usr/src/squid-6.12-1 -Wno-error=deprecated-declarations -c -o pconn.o pconn.cc
pconn.cc: In member function ‘void IdleConnList::push(const Comm::ConnectionPointer&)’:
pconn.cc:170:15: error: comparison of integer expressions of different signedness: ‘int’ and ‘size_t’ {aka ‘long unsigned int’} [-Werror=sign-compare]
  170 |     if (size_ == capacity_) {
      |         ~~~~~~^~~~~~~~~~~~
cc1plus: all warnings being treated as errors
make[3]: *** [Makefile:5929: pconn.o] Error 1
make[3]: Leaving directory '/var/cache/build/squid/squid-6.12/src'
make[2]: *** [Makefile:6046: all-recursive] Error 1
make[2]: Leaving directory '/var/cache/build/squid/squid-6.12/src'
make[1]: *** [Makefile:5039: all] Error 2
make[1]: Leaving directory '/var/cache/build/squid/squid-6.12/src'
make: *** [Makefile:600: all-recursive] Error 1

```

Reply:
My current recommendation stands: If possible, undo all pconn-related changes (rather than some of them), so that you can make progress here without rewriting that code. If not possible, let's discuss why it is not possible to revert all pconn-related changes that are seemingly unrelated to the original error.

Reply:
done

Reply:
AFAICT, this change request was resolved because problematic pconn changes were removed. The outcome of that removal experiment suggests that some pconn changes are needed. It remains to be seen whether those future pconn changes resurrect this change request. There are relevant [comments](https://github.com/squid-cache/squid/pull/1929#issuecomment-2453755443) in a discussion outside this resolved change request.